lib/platform/linux-uart.c

```diff
@@ -0,0 +1,432 @@
+		return ret;
+		
+	return 0;
+}
+
+#ifdef __CHECKER__
+#define __force __attribute__((force))
+#else
+#define __force
+#endif
+
+static void uart_close(struct switchtec_dev *dev)
+{
+	struct switchtec_uart *udev =  to_switchtec_uart(dev);
+	cli_control(dev, echo_on);
+
+	if (dev->gas_map)
+		munmap((void __force *)dev->gas_map, dev->gas_map_size);
+
+	flock(udev->fd, LOCK_UN);
+	close(udev->fd);
+	free(udev);
+}
+
+static int map_gas(struct switchtec_dev *dev)
+{
+	void *addr;
+	dev->gas_map_size = 4 << 20;
+
+	addr = mmap(NULL, dev->gas_map_size, PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+	if (addr == MAP_FAILED)
+		return -1;
+
+	dev->gas_map = (gasptr_t __force)addr;
+
+	return 0;
+}
+
+#undef __force
+
+static gasptr_t uart_gas_map(struct switchtec_dev *dev, int writeable, size_t *map_size)
+{
+	if (map_size)
+		*map_size = dev->gas_map_size;
+
+	return dev->gas_map;
+}
+
+static void uart_memcpy_from_gas(struct switchtec_dev *dev, void *dest, const void __gas *src, size_t n)
+{
+	int ret;
+	int raddr, rnum, rcrc;
+	int i, j;
+	char *pos;
+	struct switchtec_uart *udev = to_switchtec_uart(dev);
+	uint32_t addr = (uint32_t)(src - (void __gas *)dev->gas_map);
+	char gas_rd_rtn[102400];
+	uint8_t buf[n+sizeof(addr)];
+	uint8_t *ptr = buf;
+	uint8_t cal;
+
+	for (i = 0; i< RETRY_NUM; i++) {
+		ret =  send_cmd(udev->fd, NULL, SWITCHTEC_UART_GAS_READ, addr, n, src, 0);
+		if (ret)
+			raise(SIGBUS);
+
+		ret = read_resp_line(udev->fd, gas_rd_rtn);
+		if (ret)
+			raise(SIGBUS);
+
+		if(sscanf(gas_rd_rtn, "%*[^<]<0x%x> [%d Byte]%*[^:]: 0x%x%*[^:]:",
+				  &raddr, &rnum, &rcrc) != 3)
+			raise(SIGBUS);
+
+		if ((raddr != addr) || (rnum != n))
+			raise(SIGBUS);
+
+		addr = htobe32(addr);
+		memcpy(ptr, &addr, sizeof(addr));
+		ptr += sizeof(addr);
+
+		pos = strchr(gas_rd_rtn, ']');
+		pos += 2;
+		for (j = 0; j < n; j++)
+			*ptr++ = strtol(pos, &pos, 16);
+
+		cal = crc8(buf, sizeof(addr) + n, 0, true);
+		if (cal == rcrc){
+			memcpy(dest, buf + sizeof(addr), n);
+			break;
+		}
+	}
+
+	if (i == RETRY_NUM)
+		raise(SIGBUS);
+}
+
+#define create_gas_read(type, suffix) \
+	static type uart_gas_read ## suffix(struct switchtec_dev *dev, \
+		type __gas *addr) \
+	{ \
+		type ret; \
+		uart_memcpy_from_gas(dev, &ret, addr, sizeof(ret)); \
+		return ret; \
+	}
+create_gas_read(uint8_t, 8);
+create_gas_read(uint16_t, 16);
+create_gas_read(uint32_t, 32);
+create_gas_read(uint64_t, 64);
+
+static void uart_gas_write(struct switchtec_dev *dev, void __gas *dest,
+			   const void *src, size_t n)
+{
+	int ret;
+	int i;
+	char gas_wr_rtn[102400];
+	uint32_t crc;
+	uint32_t cal, exp;
+	struct switchtec_uart *udev =  to_switchtec_uart(dev);
+	uint32_t addr = (uint32_t)(dest - (void __gas *)dev->gas_map);
+	uint8_t buf[n+sizeof(addr)];
+
+	addr = htobe32(addr);
+	memcpy(buf, &addr, sizeof(addr));
+	for (i = 0; i< n; i++){
+		memcpy(buf + sizeof(addr) + i, (uint8_t *)src + n - 1 - i, sizeof(uint8_t));
+	}
+	crc = crc8(buf, sizeof(addr)+n, 0, true);
+
```

Comment:
Why is the CRC calculated in reverse data order?

Probably no sense using memcpy() to copy a single byte.  `buf[i] = ` makes more sense.

I think the best solution (if reverse CRC is a thing with this hardware) would be to write a crc8r function which scans the buffer in reverse. Then you can do:

```c
crc = crc8(&addr, sizeof(addr), 0, true);
crc = crc8r(src, n, crc, false)
```

Reply:
for gaswr -c -s <addr> <data string> <crc>, the crc calculation is protocol between utility and fw, namely start from MSB to LSB of both addr and data string.

reverse version crc8 is not needed, such as:
crc = crc8(&addr, sizeof(addr), 0, true);
for (i=n; i>0; i--)
       crc = crc8(src +i -1, sizeof(uint8_t), crc, false);

Reply:
for gasrd -c -s <addr> <n>
the crc is calculated by addr from MSB to LSB, and from data string from LSB to MSB.
it its fw implementation.